src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -57,36 +60,56 @@ public function validate($entity, Constraint $constraint)
             throw new ConstraintDefinitionException('At least one field has to be specified.');
         }
 
-        if (null === $entity) {
+        if (null === $object) {
             return;
         }
 
+        $entityClass = \get_class($object);
+
         if ($constraint->em) {
             $em = $this->registry->getManager($constraint->em);
 
             if (!$em) {
                 throw new ConstraintDefinitionException(sprintf('Object manager "%s" does not exist.', $constraint->em));
             }
         } else {
-            $em = $this->registry->getManagerForClass(\get_class($entity));
+            $em = $this->registry->getManagerForClass($constraint->entityClass ?? $entityClass);
 
             if (!$em) {
-                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($entity)));
+                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($object)));
             }
         }
 
-        $class = $em->getClassMetadata(\get_class($entity));
+        try {
+            $repository = $em->getRepository($entityClass);
+            $isEntity = true;
+        } catch (ORMMappingException | PersistenceMappingException $e) {
+            $isEntity = false;
+        }
 
-        $criteria = [];
-        $hasNullValue = false;
+        if (null !== $constraint->entityClass) {
+            /* Retrieve repository from given entity name.
+             * We ensure the retrieved repository can handle the entity
+             * by checking the entity is the same, or subclass of the supported entity.
+             */
+            $repository = $em->getRepository($constraint->entityClass);
+            $supportedClass = $repository->getClassName();
 
-        foreach ($fields as $fieldName) {
-            if (!$class->hasField($fieldName) && !$class->hasAssociation($fieldName)) {
-                throw new ConstraintDefinitionException(sprintf('The field "%s" is not mapped by Doctrine, so it cannot be validated for uniqueness.', $fieldName));
+            if ($isEntity && !$object instanceof $supportedClass) {
```

Comment:
Can you explain why you think this is needed?

Reply:
The code was originally added because the UniqueEntity constraint could fail when using inheritance  https://github.com/symfony/symfony/issues/4087
The `$constraint->entityClass` in many cases will be still used that way.
In my opinion, the message is useful when someone will make the mistake of providing an incoorect base class of validated entity class.
In a situation when object being validated is not an entity this message is useless.